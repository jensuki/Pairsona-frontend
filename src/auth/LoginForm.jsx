import { useState, useContext } from 'react';
import { useNavigate, Link } from 'react-router-dom';
import UserContext from '../context/UserContext';
import Loading from '../components/Loading';
import '../styles/Forms.css';

const LoginForm = () => {
    const { login } = useContext(UserContext);
    const navigate = useNavigate();

    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(false);
    const [formData, setFormData] = useState({
        username: '',
        password: ''
    });

    // handle form input change
    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData((fData) => ({
            ...fData,
            [name]: value
        }));
    }

    // handle form submission
    const handleSubmit = async (e) => {
        e.preventDefault();
        setError(null);
        setLoading(true); // spinner

        try {
            await login(formData);
            navigate('/'); // redirect to home
        } catch (err) {
            setError(err.message)
            setLoading(false)
        }
    }

    if (loading) return <Loading />

    return (
        <div className="FormContainer LoginForm">
            <h1>Log In</h1>
            <form onSubmit={handleSubmit}
                autoComplete="off">

                {/* username input */}
                <div className="form-group">
                    <label htmlFor="username">Username</label>
                    <input
                        type="text"
                        name="username"
                        id="username"
                        value={formData.username}
                        onChange={handleChange}
                        placeholder="Enter your username"
                        autoComplete="current-username"
                        required />
                </div>

                {/* password input */}
                <div className="form-group">
                    <label htmlFor="password">Password</label>
                    <input
                        type="password"
                        name="password"
                        id="password"
                        value={formData.password}
                        onChange={handleChange}
                        placeholder="Enter your password"
                        autoComplete="current-password"
                        required />
                </div>

                {/* error msg if exists */}
                {error && <p className="form-error">{error}</p>}
                <button type="submit" className="login-btn">Log In</button>
            </form>

            {/* sign up link */}
            <p className="signup-link">
                New user? <Link to="/signup">Create an account</Link>
            </p>
        </div>
    )
}

export default LoginForm;