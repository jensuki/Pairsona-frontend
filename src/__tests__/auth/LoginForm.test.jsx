import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import LoginForm from '../../auth/LoginForm';
import { describe, test, expect, vi, beforeEach } from 'vitest';

// mock the login function
const mockLogin = vi.fn();

describe('LoginForm component', () => {

    beforeEach(() => {
        mockLogin.mockClear();
    });

    test('renders without crashing', () => {
        render(
            <MemoryRouter>
                <UserProvider login={mockLogin}>
                    <LoginForm />
                </UserProvider>
            </MemoryRouter>
        );

        expect(screen.getByLabelText(/Username/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/Password/i)).toBeInTheDocument();
    });

    test('matches snapshot', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UserProvider login={mockLogin}>
                    <LoginForm />
                </UserProvider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('handles user input correctly', () => {
        render(
            <MemoryRouter>
                <UserProvider login={mockLogin}>
                    <LoginForm />
                </UserProvider>
            </MemoryRouter>
        );
        // enter credentials
        const usernameInput = screen.getByLabelText(/Username/i);
        const passwordInput = screen.getByLabelText(/Password/i);

        fireEvent.change(usernameInput, { target: { value: 'testuser' } });
        fireEvent.change(passwordInput, { target: { value: 'password123' } });

        expect(usernameInput.value).toBe('testuser');
        expect(passwordInput.value).toBe('password123');
    });

    test('calls login func when submitted', async () => {
        mockLogin.mockResolvedValue(); // simulate success

        render(
            <MemoryRouter>
                <UserProvider login={mockLogin}>
                    <LoginForm />
                </UserProvider>
            </MemoryRouter>
        );
        // enter credentials and click login
        const usernameInput = screen.getByLabelText(/Username/i);
        const passwordInput = screen.getByLabelText(/Password/i);
        const loginBtn = screen.getByRole('button', { name: /Log In/i });

        fireEvent.change(usernameInput, { target: { value: 'testuser' } });
        fireEvent.change(passwordInput, { target: { value: 'password123' } });
        fireEvent.click(loginBtn);

        await waitFor(() => {
            expect(mockLogin).toHaveBeenCalledWith({
                username: 'testuser',
                password: 'password123',
            });
        });
    });

    test('displays error message for invalid credentials', async () => {
        // simulate failure with error msg
        mockLogin.mockRejectedValue(new Error('Invalid username or password'));

        render(
            <MemoryRouter>
                <UserProvider login={mockLogin}>
                    <LoginForm />
                </UserProvider>
            </MemoryRouter>
        );

        const usernameInput = screen.getByLabelText(/Username/i);
        const passwordInput = screen.getByLabelText(/Password/i);
        const loginBtn = screen.getByRole('button', { name: /Log In/i });

        fireEvent.change(usernameInput, { target: { value: 'wronguser' } });
        fireEvent.change(passwordInput, { target: { value: 'wrongpassword' } });
        fireEvent.click(loginBtn);

        // expect proper error
        const errorMsg = await screen.findByText(/Invalid username or password/i);
        expect(errorMsg).toBeInTheDocument();
        expect(errorMsg).toHaveClass('form-error');
    });
});
