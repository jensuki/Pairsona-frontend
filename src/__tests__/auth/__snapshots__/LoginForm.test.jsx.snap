// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FormContainer LoginForm"
  >
    <h1>
      Log In
    </h1>
    <form
      autocomplete="off"
    >
      <div
        class="form-group"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="current-username"
          id="username"
          name="username"
          placeholder="Enter your username"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          id="password"
          name="password"
          placeholder="Enter your password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        class="login-btn"
        type="submit"
      >
        Log In
      </button>
    </form>
    <p
      class="signup-link"
    >
      New user? 
      <a
        data-discover="true"
        href="/signup"
      >
        Create an account
      </a>
    </p>
  </div>
</DocumentFragment>
`;
